### ISSUE TITLE
Contributors Request

### ISSUE THREAD

**Issue by CarterLi** [CarterLi/iina#4]
As IINA is my favorite macOS player, I really hope that IINA can go further. However the official repo seemed to be no longer in active development ( but not dead, hopefully ).

Since then, many features, fixes were made in various personal fork repo. However because they were separated in different repos:

1. We cannot release binaries that includes all of these features / fixes
1. Other IINA users are hard to find them

So I decided to create this [iina-plus](https://github.com/iina-plus) group and want it to be open and easy to contribute ( not open PR, but join the group and push code directly into the repo )

However, to prevent from spam, there is a requirement to prove you have the ability to contribute.

* You must have at least one existing working feature / important bug fix in your repo.

**Comment by CarterLi** [comment]
@low-batt I really appreciate your PRs with very detailed explanation and most of them have been cherry-picked into this repo. Would you like to join this group?

**Comment by low-batt** [comment]
I tried forking iina-plus/iina today as I like the ability to look over your submission on GitHub the PR workflow gives you before you push it into the public repository. Yes I'm paranoid about accidentally messing up and pushing something I didn't intend. I found I couldn't fork ina-plus/iina. Seemed like this is a restriction on GitHub free accounts.

So yes, if you grant me access I will add my changes to this fork as well.

What is your vision for coordinating changes given this setup?

**Comment by low-batt** [comment]
I have a local commit with the fix for the camera housing issue but:
low-batt@gag iina (develop>)$ git push
ERROR: Permission to iina-plus/iina.git denied to low-batt.

**Comment by CarterLi** [comment]
Let me have a check

**Comment by CarterLi** [comment]
Try again please?

**Comment by low-batt** [comment]
Worked!
I haven't had a chance to look through what all you picked up from me.
Are there other commits of mine you want me to add?

**Comment by CarterLi** [comment]
Some commits you made had conflict with my repo. I skipped them. You may have a check.

**Comment by low-batt** [comment]
I will check into it and where appropriate adjust the commits and get them pushed into this fork.
May take me a while.

**Comment by low-batt** [comment]
Pushed a commit with the fix for https://github.com/iina/iina/issues/3574

**Comment by low-batt** [comment]
I have just submitted a pull request with a fix for issue https://github.com/iina/iina/issues/3475. I was unable to reproduce the crash, so I can't confirm it will fix the crashes. Do you want to wait for the core developers to review the fix, or do you want me to push it to iina-plus?

**Comment by CarterLi** [comment]
iina-plus is a repo that experiments new features and bug fixes that haven't been officially accepted by upstream.

It's ok to push your changes directly into iina-plus repo and we may revert it later if necessary. Let's release a new binary for users to test the changes.

**Comment by low-batt** [comment]
Pushed fix for https://github.com/iina/iina/issues/3475

**Comment by low-batt** [comment]
Pushed fix for https://github.com/iina/iina/issues/3584

**Comment by low-batt** [comment]
Pushed fix for https://github.com/iina/iina/issues/3590

Going to have to keep an eye on this one. The fix is correct and needed. But because IINA is now waiting for mpv to shutdown we may find issues in the way mpv is being shutdown that usually were hidden by premature termination of the application. I sometimes see log messages from mpv that makes me think they might have some bugs in their quit command.

**Comment by low-batt** [comment]
Pushed fix for https://github.com/iina/iina/issues/3551

**Comment by low-batt** [comment]
Pushed fix for iina#3593

This fixes an existing issue in the shutdown sequence that was made more obvious by the fix to properly wait for mpv to terminate.

**Comment by CarterLi** [comment]
Thanks.

If your PR gets merged, I'd like to rebase everything on top of the official branch. Please don't be surprised if you get any conflicts.

**Comment by ViRo3** [comment]
Here's a video for the title bar issue. Will be adding logs for the crash/freeze

https://user-images.githubusercontent.com/87166847/147728513-79009097-51c7-499d-add4-642384dcb1bd.mp4

**Comment by low-batt** [comment]
Pushed a fix to the fix for https://github.com/iina/iina/issues/3558
Simple issue with not closing the mask window if IINA exits directly from full screen.
I'm ready to help with resolving any merge conflicts.
I'm expecting it will take a while for them to review all my changes and get them merged.

**Comment by low-batt** [comment]
Pushed a commit to update the copyright year to 2022

**Comment by low-batt** [comment]
Pushed fix for https://github.com/iina/iina/issues/3601

**Comment by low-batt** [comment]
Pushed a tiny fix for iina#3607

**Comment by low-batt** [comment]
Pushed small fix for iina#3211

**Comment by CarterLi** [comment]
Just rebased everything onto the official repo

**Comment by low-batt** [comment]
Pushed a revision to the fix for iina#3558

**Comment by CarterLi** [comment]
> Pushed a revision to the fix for [iina#3558](https://github.com/iina/iina/issues/3558)

Since you have closed the old PR in the official repo, please consider amending your addition fixes into the old one, to avoid possible rebase conflicts. @low-batt

**Comment by low-batt** [comment]
That is beyond my knowledge of git. How would I do that?
I've only amended commits that had not been merged.

**Comment by CarterLi** [comment]
Just remove the [old commit](https://github.com/iina-plus/iina/commit/35337e7bc0d99b55a95c02f326b67639abd22afa) using `git rebase -i` then commit [your new changes](https://github.com/iina/iina/pull/3618/commits/9f110afbfe0654e61782565da20a0ffdceb05e30).

Force-pushing is needed. It's dangerous but necessary to maintain a fork repo

**Comment by low-batt** [comment]
There are 3 commits related to the camera I had to drop. I had to go way back to get that first one. This is what the rebase looks like. Wanted to make sure you think this is ok before I force push.
```text
low-batt@gag iina (develop $=)$ git rebase -i HEAD~55
Auto-merging iina/Info.plist
CONFLICT (content): Merge conflict in iina/Info.plist
Auto-merging iina.xcodeproj/project.pbxproj
error: could not apply 9d6d4139... Add AppleScript support
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 9d6d4139... Add AppleScript support
low-batt@gag iina (develop *+$%|REBASE 45/55)$ git status
interactive rebase in progress; onto b6e6dedb
Last commands done (45 commands done):
   pick a2900c3f Backout fix for CPU is consumed by OSC while hidden iina#3601
   pick 9d6d4139 Add AppleScript support
  (see more in file .git/rebase-merge/done)
Next commands to do (10 remaining commands):
   pick f0a99d53 Revert `MACOSX_DEPLOYMENT_TARGET` back to `10.12`
   pick a785a18a Update README and ISSUE_TEMPLATE
  (use "git rebase --edit-todo" to view and edit)
You are currently rebasing branch 'develop' on 'b6e6dedb'.
  (fix conflicts and then run "git rebase --continue")
  (use "git rebase --skip" to skip this patch)
  (use "git rebase --abort" to check out the original branch)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   iina.xcodeproj/project.pbxproj
	new file:   iina/IINA.sdef
	modified:   iina/MPVPlaylistItem.swift
	modified:   iina/MPVTrack.swift
	modified:   iina/PlayerCore.swift
	modified:   iina/PrefGeneralViewController.swift
	new file:   iina/Scripting/AppDelegate+Scripting.swift
	new file:   iina/Scripting/Commands/NextCommand.swift
	new file:   iina/Scripting/Commands/NextFrameCommand.swift
	new file:   iina/Scripting/Commands/PauseCommand.swift
	new file:   iina/Scripting/Commands/PlayCommand.swift
	new file:   iina/Scripting/Commands/PlayPauseCommand.swift
	new file:   iina/Scripting/Commands/PreviousCommand.swift
	new file:   iina/Scripting/Commands/PreviousFrameCommand.swift
	new file:   iina/Scripting/FourCharCode+Extensions.swift
	new file:   iina/Scripting/MPVPlaylistItem+Scripting.swift
	new file:   iina/Scripting/MPVTrack+Scripting.swift
	new file:   iina/Scripting/PlayerCore+Scripting.swift

Unmerged paths:
  (use "git restore --staged <file>..." to unstage)
  (use "git add <file>..." to mark resolution)
	both modified:   iina/Info.plist

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	deps/include/libavcodec/defs.h
	deps/include/libavdevice/
	deps/include/libavfilter/
	deps/include/libavutil/detection_bbox.h
	deps/include/libpostproc/
	deps/include/libswresample/

low-batt@gag iina (develop *+$%|REBASE 45/55)$ 
low-batt@gag iina (develop +$%|REBASE 45/55)$ git rebase --continue
[detached HEAD b8f271c3] Add AppleScript support
 Author: 李通洲 <carter.li@eoitek.com>
 19 files changed, 1150 insertions(+), 3 deletions(-)
 create mode 100644 iina/IINA.sdef
 create mode 100644 iina/Scripting/AppDelegate+Scripting.swift
 create mode 100644 iina/Scripting/Commands/NextCommand.swift
 create mode 100644 iina/Scripting/Commands/NextFrameCommand.swift
 create mode 100644 iina/Scripting/Commands/PauseCommand.swift
 create mode 100644 iina/Scripting/Commands/PlayCommand.swift
 create mode 100644 iina/Scripting/Commands/PlayPauseCommand.swift
 create mode 100644 iina/Scripting/Commands/PreviousCommand.swift
 create mode 100644 iina/Scripting/Commands/PreviousFrameCommand.swift
 create mode 100644 iina/Scripting/FourCharCode+Extensions.swift
 create mode 100644 iina/Scripting/MPVPlaylistItem+Scripting.swift
 create mode 100644 iina/Scripting/MPVTrack+Scripting.swift
 create mode 100644 iina/Scripting/PlayerCore+Scripting.swift
Successfully rebased and updated refs/heads/develop.
low-batt@gag iina (develop $<>)$ git status
On branch develop
Your branch and 'origin/develop' have diverged,
and have 51 and 54 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

nothing to commit, working tree clean
low-batt@gag iina (develop $<>)$ 
```

**Comment by CarterLi** [comment]
It's hard to tell whether it's ok or not. Just make sure that it builds without errors and go on

**Comment by low-batt** [comment]
Was getting sleepy, so held off until brain was awake again. I confirmed build worked. I compared the git log taken before and after showed only the 3 camera housing fix related commits were removed. I have now force pushed. I always find that operation scary.

I have pushed the replacement fix for the camera housing as well as two additional commits coming out of code review for iina#3334, problems with taking a screenshot.

**Comment by CarterLi** [comment]
I will have a holiday for about 2 weeks. During the holiday, I will have no access to my mac. Please release new builds yourself if necessary @low-batt

**Comment by low-batt** [comment]
Relax and have a great time on vacation. :-)

Not sure what the procedure is for creating a release. I've never done that on GitHub. I will try and figure that out.

I was going to be asking you for help on these problems. Seeing some really strange behavior in the logs that I can't explain.

**Comment by low-batt** [comment]
Pushed fix for iina#3620

I've been abusing this issue and using it for communication.
Thoughts on enabling GitHub discussions for this repo?

**Comment by CarterLi** [comment]
> Pushed fix for [iina#3620](https://github.com/iina/iina/issues/3620)
> 
> I've been abusing this issue and using it for communication. Thoughts on enabling GitHub discussions for this repo?

I was enabling GitHub discussion. However I found that I always forgot to check it so I closed it.

**Comment by low-batt** [comment]
Pushed a commit that tweaks the app and document icons to match up with the latest HIG requirements.
Thank you to @Zabriskije who provided the updated icons.
More details can be found in discussion https://github.com/iina/iina/discussions/3670.

**Comment by low-batt** [comment]
Pushed fix for https://github.com/iina/iina/issues/3679

IINA already supports playing animated GIFs, but was not associated with .gif files.

**Comment by low-batt** [comment]
Pushed fix for https://github.com/iina/iina/issues/3692

**Comment by low-batt** [comment]
Pushed fix for https://github.com/iina/iina/issues/3695

**Comment by CarterLi** [comment]
With the latest upstream changes, I got conflicts when rebasing. Can you please resolve it?

**Comment by low-batt** [comment]
Yes, I will take care of rebasing, probably later today. Reviewing and merging is currently happening in the main IINA repository. I'm about to respond to review feedback.

**Comment by low-batt** [comment]
Pushed commit that adds OSD messages for file loop and playlist loop, https://github.com/iina/iina/issues/3229

There is other cleanup that might be needed for IINA+. There are commits that were added for debugging when trying to figure out the hang that turned out to be an issue with hardware drivers and FLAC encoded files. I will check over the commits and get IINA+ aligned with the official repository.

**Comment by low-batt** [comment]
Resolved rebase conflicts. Up to date with upstream for now.
They are reviewing PRs, so more merging may be required soon.

**Comment by low-batt** [comment]
Pushed fix for https://github.com/iina/iina/issues/3462

**Comment by low-batt** [comment]
More commits have been merged, so IINA+ is behind again. Likely these will trigger merge conflicts as well.
I will look into rebasing again tomorrow.

**Comment by low-batt** [comment]
Starting rebase again. May take a while as I expect merge conflicts with my changes.

@CarterLi Did you notice the request that you post a HDR PR in the official repository?
I told the devs to hold off merging my changes to  FFmpegController in PR https://github.com/iina/iina/pull/3461 as I know there are lots of changes to that source for HDR. Better to get HDR in first and let me deal with merge conflicts.

**Comment by low-batt** [comment]
Completed rebase.
Noticed HDR PR posted.
😄

**Comment by low-batt** [comment]
Starting rebase with IINA again.

**Comment by low-batt** [comment]
Completed rebase with IINA.

**Comment by CarterLi** [comment]
Did a major rebase

**Comment by low-batt** [comment]
I pulled the latest IINA+ and rebuilt.

Starting with both `File Loop` and `Playlist Loop` enabled in defaults.
I bring up IINA and both are now correctly checked in the `Playback` menu.
So that is fixed.
I disable `Playlist Loop` using the menu item.
I quit IINA.
I check defaults and `Playlist Loop` is still enabled in the saved defaults.

I also experienced another inconsistency.
Starting again with both enabled in defaults.
I brought up IINA.
Menu correctly shows both enabled.
Opening playlist, the `Playlist Loop` button shows enabled, but the `File Loop` button is in the disabled state.

On another topic, I am actively working on the issues with the display refresh rate matching feature.

**Comment by low-batt** [comment]
Starting rebase with IINA again.

**Comment by low-batt** [comment]
Completed rebase with IINA.

**Comment by low-batt** [comment]
Rebased with IINA again.

**Comment by CarterLi** [comment]
It's hard to maintain a fork. It's even harder to maintain an Xcode project fork with a lot of IDE-generated code.

The upstream commit <https://github.com/iina/iina/commit/d37734b72a58558ee76a05e34d5050e7385c5982> introduced major conflicts against this repo. I have tried to resolve the conflicts and I failed. And even if I manage resolving the conflicts, the upstream branch <https://github.com/iina/iina/tree/plugin-system> will introduce much more conflicts and I don't think I can resolve them.

As the major feature, HDR playback, has been merged into upstream, I truly doubt if it's worth maintaining this fork...

**Comment by low-batt** [comment]
Pushed another fix for https://github.com/iina/iina/issues/3590

This fix is compatible with the previous fix. The new fix addresses crashes in the logger directly.

**Comment by low-batt** [comment]
@CarterLi, The IINA project is reviewing and merging and asking about HDR fixes.
Can you post PRs for?:
Issue https://github.com/iina/iina/issues/3806, the random behavior of enabling.
Issue https://github.com/iina/iina/issues/3808, a preference to disable HDR.
Are there other HDR fixes that should be pushed to the IINA repo?

**Comment by CarterLi** [comment]
No. There are a lot of conflicts between my repo and official one. I don't bother resolve them.

**Comment by low-batt** [comment]
How do we get those issues fixed in the official repo? Do you want me to try and figure out what changes are needed?

**Comment by CarterLi** [comment]
All the issues you mentioned above were linked with my changes. All HDR related commits were prefixed with `HDR:`

**Comment by low-batt** [comment]
Found the commits. I will get those two issues fixed in the main repo.

I've not had time to try a rebase and see how bad the conflicts are. I can easily deal with conflicts in the changes I've made. I've been worried all those changes for the plugin system would prove challenging to deal with. I've not had a big block of time to look into this.

**Comment by low-batt** [comment]
I've just upgraded to the publicly released Ventura and that strange dimming that only affects HDR in full screen mode is still there. This is issue https://github.com/iina/iina/issues/3880 reported against the Ventura Beta that we decided must be a macOS defect. After reproducing it and looking into it I don't see how it can be anything other than a defect. But Apple has let it out to the public and more users are now reporting it. This is not slight dimming. In the test I'm running the picture is nearly black. Definitely unwatchable.

I just tested a truly horrible awful hackaround. If I add a tiny (0.1 x 0.1) nearly opaque (0.01) view as a subview then the screen does not dim. I really hate this workaround.

Before proposing such an awful workaround I wanted to check in and see if you had any other ideas about this problem or better ways to workaround it.

**Comment by CarterLi** [comment]
I haven't had time to look into it. Will investigate it later

**Comment by low-batt** [comment]
IINA ended up going with the awful workaround. It is in PR https://github.com/iina/iina/pull/4031. Do you want me to add that to IINA+?

mpv just released [v0.35.0](https://github.com/mpv-player/mpv/releases/tag/v0.35.0). I haven't had a chance to try it yet.

**Comment by CarterLi** [comment]
No. I don't like the hack.

**Comment by low-batt** [comment]
Yes, it is a horrible hackaround. I will not put it in IINA+. Ventura definitely has some odd graphics defects. It has been reported that 13.0.1 is no better. Some of these problems have been reproduced without IINA, so it is definitely macOS that is at fault.
